modules/mbedtls_curl/http_client_curl.h

```diff
@@ -0,0 +1,128 @@
+/*************************************************************************/
+/*  http_client_curl.h                                                   */
+/*************************************************************************/
+/*                       This file is part of:                           */
+/*                           GODOT ENGINE                                */
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2021 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2021 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef HTTP_CLIENT_CURL_H
+#define HTTP_CLIENT_CURL_H
+
+#include "core/io/http_client.h"
+#include "core/templates/ring_buffer.h"
+
+#include <curl/curl.h>
+#include <stdio.h>
+
+struct RequestContext {
+	RequestContext(){};
+	~RequestContext();
+
+	List<String> *response_headers = nullptr;
+	int *response_code = nullptr;
+	RingBuffer<uint8_t> *read_buffer = nullptr;
+	curl_slist *header_list = nullptr;
+	int *body_size = nullptr;
+	HTTPClient::Status *status = nullptr;
+	Vector<PackedByteArray> *response_chunks = nullptr;
+	bool *has_response = nullptr;
+	bool *chunked = nullptr;
+	bool *keep_alive = nullptr;
+};
```

Comment:
Is there any reason to use this `RequestContext` with all pointers instead of just passing the `HTTPClientCurl` itself to (i.e. `this`) to the callbacks?
If we go for the `RequestContext` approach (which could be also good for future-proofing), those members should not be pointers, but the actual types.
We can add a `RequestContext current_context` member to `HTTPClientCurl` and pass the pointer to that to the callbacks (and add a `RequestContext::clear` for clearing it).

Reply:
Okay @Faless , I've updated the PR to remove the `RequestContext` struct entirely. It not stores the request state in the HttpClient object itself. I remember now that the reason I implemented the RequestContext was because I wasn't sure whether or not there could be multiple requests for the same client in flight at once. It doesn't seem like there can be given the way the interface is designed.